4750 angular 10 tour

mrsarm opened the pull request:
# Description

Ticket: medic/cht-core#4750

Migration of the guide tour components.

### Notes

- Because some of the tabs were not migrated yet, in those tabs the tour does not work, so once migrated a new PR may be needed to adapt the tour configuration to the new components (tough should be a straightforward PR). Now it works in the _"Reports"_, _"Analytics"_ and the _"Messages"_ tabs.
- As discussed in a meeting, the bootstrap-tour component was removed because it depends of Bootstrap JS (Popover), and used instead bootstrap-tour-standalone, but because it is not available in the Npm registry, I have added the source code of the JS and the CSS module (BTW, I put it in the root of the CSS and TS folders the files, not sure where would be the best path to place the files).

# Code review items

- Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

dianabarsan reviewed (changes requested):
Nice work, thanks! 

I left a couple of comments inline.
Besides those, I've also noticed some problems with the tour, compared to master: 

1. Tabs have text on mobile
![image](https://user-images.githubusercontent.com/35681649/98809155-ca382880-2425-11eb-9ccb-6cd104a7b4c0.png)

2. If I have a tour active, and I navigate to another tab, the tour is stuck on the screen. and doesn't respond to commands: 

![image](https://user-images.githubusercontent.com/35681649/98809264-f0f65f00-2425-11eb-94d9-e6d246b04a46.png)

3. In mobile view, as I follow the reports tour, at step 8 (?) where the app is supposed to navigate to the content page, the tour just doesn't do anything. 
On `master` it navigates to the `content` page, but the tour is lost. 
Have you checked if there is an open issue for this? 
Please open issues for bugs that you also observe in `master` during your migration!

mrsarm commented on webapp/src/ts/modals/tour/tour-select.component.ts:
```diff
@@ -0,0 +1,39 @@
+
+  ngOnInit() {
+    this.tourService.getTours().then(tours => {
+      this.tours = tours.sort((a: any, b: any) => a.order - b.order);
```
I agree that should be returned sorted b the service, I did in the other way just to replicate the old behaviour, where it was sorted in the component, but in the template [here](https://github.com/medic/cht-core/blob/d9ae1aa/webapp/src/templates/modals/tour_select.html#L8):

```html
<a ng-repeat="tour in tourSelectCtrl.tours | orderBy:'order' track by tour.id" class="...
```

Because Angular 2+ does not support sorting in the templates any more, I moved the sorting to the component, but I can move it to the service.

mrsarm commented on webapp/src/ts/services/startup-modals.service.ts:
```diff
@@ -0,0 +1,202 @@
+      required: (settings, user) => !user.known && this.tours.length > 0,
+      render: () => {
+        this.openTourSelect();
+        return this.updateUserService.update(
```
Yes the modal is opened without returning any promise because it called passing as argument the component class, not an instance, that is way in the function the execution is not suspended waiting for the return from the modal, but I see that the [ModalService.show()](https://github.com/medic/cht-core/blob/07ad95a5e4641afaa52310780c9162e2d297f563/webapp/src/ts/modals/mm-modal/mm-modal.ts#L81) method can be called passing a handler that is subscribed to be called when the modal is closed, right? I'll take a look...

mrsarm commented on webapp/src/ts/services/startup-modals.service.ts:
```diff
@@ -0,0 +1,202 @@
+        //   controllerAs: 'welcomeModalCtrl',
+        //   size: 'lg',
+        // }).catch(() => {});
+        return Promise.resolve();
```
Will try, if I don't have to migrate a other components, otherwise may be better to make it in another PR.

mrsarm commented on webapp/src/ts/bootstrap-tour-standalone.js:
```diff
@@ -0,0 +1,1703 @@
+/* ========================================================================
```
OK, I though we were going to remove the `webapp/src/js` folder at some point.

dianabarsan replied:
We are, at some point, but we still have js files that we're not migrating this iteration.

mrsarm commented on webapp/tests/karma/ts/services/startup-modals.service.spec.ts:
```diff
@@ -0,0 +1,135 @@
+    sinon.restore();
+  });
+
+  function getProviders(toursResult) {
```
Yes the problem is that I in both tests the configuration is almost the same, except for the one bean that in its constructor executes a call to another service that is mocked, and the execution happens only one time at the beginning of the initialization, so I can refactor like you said to make 2 `beforeEach` functions, but each one will only be used by one test case.

dianabarsan replied:
I think you can change the moked result before you inject the service.

mrsarm replied:
I tried what you said in this [fork](https://github.com/medic/cht-core/blob/4750-angular-10-tour-startup-modals-spec-fix/webapp/tests/karma/ts/services/startup-modals.service.spec.ts) of the branch but the stub does not work, I got:

```
      ✖ "before each" hook for "Tour modal should not be displayed if no tours are available"
        Chrome Headless 86.0.4240.193 (Linux x86_64)
      TypeError: Cannot read property 'then' of undefined
          at new StartupModalsService (_karma_webpack_/src/ts/services/startup-modals.service.ts:71:32)
```
Like the override of the stub in the method itself is executed later of the service initialization.

I can try a double nested `describe` section, with a main `beforeEach` and a `beforeEach` on each nested `describe`, but not sure if it will work and still will only have one test on each describe/describe block.

mrsarm commented:
All the corrections were made. However as noticed in the PR comments, there are some bugs that comes before this PR from the tour guide (master version), but others were found here and were addressed. One of the errors in the current master branch and worst here is that when you click in another tab or try to launch the tour again while there is one running, the one running does not close, shadowing the new one, or staying behind. Because it is easier to fix the issue than make the bug behave exactly the old one, I fixed the issue and now the old tour is closed if a new one is launched or a tab is clicked (just for the tabs migrated, not working the fix yet if you click in the Tasks or the Contacts tabs not migrated).

Similar bugs may appear if you have the tour running and then click in another option from the menu, but are errors that come from the master version. I'll create separate issues to address those bugs later, outside the migration.

The tour is not working well in the Messages tab when using from a mobile version, but it is because tab is not working well, eg. if you click the tab, then open a message, then go to another tab, then go back to the message tab and the message previously visited is opened again instead of the list view. Not sure if this bug was noticed, here is a recording:

![Messages Mobile Error 2020-11-16](https://user-images.githubusercontent.com/1608415/99317859-b5e49900-2834-11eb-823b-4e7af6f27f6f.gif)

About the tour not working well in the mobile version of the Reports tab, I tested it after many changes and worked well, so not sure if I fixed the bug or I'm not able to reproduce it, here is a recording of how it is working now in my local environment with the latest changes:

![Reports Tour Mobile 2020-11-16](https://user-images.githubusercontent.com/1608415/99318251-59ce4480-2835-11eb-8c3e-2c35655c63a1.gif)

dianabarsan commented:
@mrsarm for what you've noticed about the messages tab, can you please create an issue in the migration branch? 
cc @latin-panda

dianabarsan reviewed (changes requested):
Really nice work! Thanks! 

I added some really small comments inline. 
I'm still seeing the problem with the reports tour not navigating to content on mobile. I can share a link to my local instance if you want to debug.

dianabarsan commented:
I'm also seeing that the tour is sticky when I navigate to a tab that doesn't have a tour while a tour is active:

![image](https://user-images.githubusercontent.com/35681649/99413881-e7299b80-28fe-11eb-97a7-57d083f74139.png)

![image](https://user-images.githubusercontent.com/35681649/99413953-fc062f00-28fe-11eb-9893-368df2e6a58d.png)

mrsarm commented on webapp/src/ts/modals/mm-modal/mm-modal.ts:
```diff
@@ -26,7 +30,15 @@ export class MmModal {
+  ngOnInit() {
+    this.store
+      .select(Selectors.getMinimalTabs)
+      .subscribe(minimalTabs => {
```
Because mm-modal is a generic approach I'll take the general approach to subscribe and then unsubscribe on destroy, despite now is only useful for the tour modal.

dianabarsan replied:
I think we should really not pollute the mm-modal abstract. If possible, please subscribe in the child modal

mrsarm commented on webapp/src/ts/modals/tour/tour-select.component.html:
```diff
@@ -0,0 +1,15 @@
+  [titleKey]="'tour.select'"
+  (onCancel)="close()"
+>
+  <div role="document" class="tour-select">
```
Like above, I took a general approach to make that CSS class available in the modal , so other components inside modals can make use of it, despite now only been used by the tour modal.

dianabarsan replied:
I'd rather we don't pollute the abstract class.

dianabarsan replied:
The class is unlikely to ever be useful to other modals because it's related to visible tabs. I sure we're not going to add more modals in the near future that will make use of it. Let's try to keep things clean, please. Thanks!

mrsarm replied:
Sorry I don't agree, but I'll. In the master version the `hide-mobile-tab-labels` class was in the `body` tag, not in a inner `div` tag. The changes in the migration moved that CSS out of the body causing modal components to not inherit the class any more. There are other modals not using the class but we don't know for sure if it could be useful in the future, because it defines how icons are shown, for now only the one at the top and the same ones in the tour, but may be useful for other workflows with icons displayed if we want to maintain the same look & feel. Moreover, this is only used if a modal is created, and once the modal is destroyed, the objects listening the variable are destroyed.

dianabarsan replied:
It defines how icons are shown depending on how many tabs the user can see in mobile view. 
If any other tab needs it, we can add it then. Until then, I'd prefer to keep the abstract class clean.

mrsarm commented on webapp/src/ts/services/tour.service.ts:
```diff
@@ -0,0 +1,598 @@
+    }
+    const tour = this.getTour(name);
+    const route = tour && tour.route;
+    setTimeout(() => {
```
I just migrated the old code trying to not refactor too much, but I'll try to remove it and check whether it doesn't break the component.

mrsarm replied:
Removed, doesn't seem to affect the startup of the tour, but must be well tested later.

mrsarm commented on webapp/tests/karma/ts/modules/messages/messages.component.spec.ts:
```diff
@@ -55,7 +59,8 @@ describe('Messages Component', () => {
           { provide: exportService, useValue: {} },
-          { provide: ModalService, useValue: modalService }
+          { provide: ModalService, useValue: modalService },
+          { provide: TourService, useValue: tourServiceMock }
```
We may remove the all tour thing as discussed in a meeting earlier, so I did migrate the existing tests including the one from the tour component but not sure if worth to add more code related with it.

mrsarm commented:
> I'm also seeing that the tour is sticky when I navigate to a tab that doesn't have a tour while a tour is active:
> 
> ![image](https://user-images.githubusercontent.com/35681649/99413881-e7299b80-28fe-11eb-97a7-57d083f74139.png)
> 
> ![image](https://user-images.githubusercontent.com/35681649/99413953-fc062f00-28fe-11eb-9893-368df2e6a58d.png)

Yes, this is an existing bug in master and it happens with all the tabs (I'll create tomorrow new tickets in the cht-core with all the bugs I found migrating the tour components here). To fix another bug related with the tour guide I fixed this one in this PR, but only in the tabs that are already migrated. Once the other tabs are migrated, the correction will be applied automatically because the way the is initialized the tour on each tab: https://github.com/medic/cht-core/blob/1727326390/webapp/src/ts/modules/reports/reports.component.ts#L117

dianabarsan commented:
> Yes, this is an existing bug in master

I'm not seeing it in `master`. I just rechecked on `gamma`. I pop the tour, select a random one (contacts), and after being redirected and the tour popping up, I open the hamburger menu and hit "about". The tour is gone on the "About" page.

dianabarsan reviewed (changes requested):
Just the inline comments about keeping the abstract modal clean and the bug with the sticky tour.

dianabarsan commented:
About the tour deactivation when navigating to another tab, this is the code that achieves it on master: 
https://github.com/medic/cht-core/blob/master/webapp/src/js/controllers/inbox.js#L310
```
$transitions.onSuccess({}, function(trans) {
      Tour.endCurrent();
      ctrl.setCurrentTab(trans.to().name.split('.')[0]);
      if (!$state.includes('reports')) {
        ctrl.setSelectMode(false);
      }
    });
```

mrsarm commented:
> > Yes, this is an existing bug in master
> 
> I'm not seeing it in `master`. I just rechecked on `gamma`. I pop the tour, select a random one (contacts), and after being redirected and the tour popping up, I open the hamburger menu and hit "about". The tour is gone on the "About" page.

I took a look to other demos like the covid and it is not working, now seeing the master version you are right, but as I said, the solution is there, you are testing tabs that are not working, once those tabs are migrated, the the solution will be applied.

dianabarsan commented:
> once those tabs are migrated, the the solution will be applied.

I tried a tab that doesn't have a tour :) The about page. There are plenty of other pages that don't disable the tour.

mrsarm commented:
> > once those tabs are migrated, the the solution will be applied.
> 
> I tried a tab that doesn't have a tour :) The about page - which is already migrated. There are plenty of other pages that don't have tours configured, and if I navigate to any of them, the existing tour is sticky.
> 
> The solution you proposed only works when navigating to tabs that _have a tour_ configured.

I agree that this require a broader solution, I just trying to just migrate without too much refactor as it was the intention, in master there are plenty of actions that does not close the running tour: clicking the hamburger menu (which is very annoying specially in the mobile version), clicking in the message tab to create a new message, clicking in the report tab filters... at this point not sure what is the best approach, try to spend more time looking for a more broad solution, or just check one by one action.

dianabarsan commented:
I think we should go for the broad solution, by finding the migrated equivalent of how it was implemented on master, the code snippet I linked in a comment above: https://github.com/medic/cht-core/blob/master/webapp/src/js/controllers/inbox.js#L310

For example, this is how we set `currentTab` now: https://github.com/medic/cht-core/blob/4750-angular-10-take-2/webapp/src/ts/app.component.ts#L132
Maybe we could close the current tour on a navigation event, just like we did in angularJS.

dianabarsan commented:
Sorry @mrsarm , I found another issue. The startup modals are triggered before accepting the privacy policy. 

![image](https://user-images.githubusercontent.com/35681649/99658451-3212f280-2a68-11eb-989e-aaa7faff3fd9.png)

The way it worked before was that the startup modals were in a separate component (directive) which was only loaded AFTER the user accepted the privacy policy. 
On master, the tour select appears only _after_ the user accepted the privacy policy. 
To see this issue, add a privacy policy for your language, create a new user and login with this user.

dianabarsan reviewed (changes requested):
The code looks great! 

The only problem is the one in the above comment about startup modals and privacy policies interaction. 

I'm sorry for not checking this earlier :( and going through this back and fourth with you. 
Thanks a lot for bearing with me on this and all the great work you've been doing!

mrsarm commented:
> The code looks great!
> 
> The only problem is the one in the above comment about startup modals and privacy policies interaction.
> 
> I'm sorry for not checking this earlier :( and going through this back and fourth with you.
> Thanks a lot for bearing with me on this and all the great work you've been doing!

Thanks Diana, no problem, I didn't notice that there is a privacy policy view because I didn't have a policy configured in my local environment so the view was not showing up in my local env, thanks to check this edge cases! please take a look again with the new changes.

dianabarsan commented on webapp/src/ts/modules/privacy-policy/privacy-policy.component.ts:
```diff
@@ -48,7 +50,7 @@ export class PrivacyPolicyComponent implements OnInit {
       .accept(this.privacyPolicy)
-      .then(() => this.globalActions.setPrivacyPolicyAccepted(true));
+      .then(() => this.globalActions.setPrivacyPolicyAccepted(true))
+      .then(() => this.startupModalsService.showStartupModals());
```
I think it was a little more elegant and clean before, when startup-modals was a component and we controlled when it was loaded from the main template and everything was handled in a single place. 

I think the privacy policy component shouldn't need to know about startup modals.

dianabarsan replied:
On second though, maybe it's not worth the extra code + rendering time to have this additional component and the direct interaction makes things more obvious.

dianabarsan replied:
coin flip? :D

mrsarm replied:
Link to the issue: https://github.com/medic/angular10-migration/issues/38

dianabarsan reviewed (approved):
Thanks for the hard work!

I have one slight objection, but I'm approving this because it's not critical and we can easily update this later, should we choose to.
